[core][autoscaler][v2] node providers - v2 adaptors [2/x]

rickyyx opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?
This PR implements a v2 node provider adaptor to work with v1 node providers.
- It also refactors the FakeMultiNodeProvider for testing 
- Modifies the CloudProviderError to be an actual exceptions so that it could be raised through call stack.

Part of https://github.com/ray-project/ray/issues/42840

<!-- Please give a short summary of the change and the problem this solves. -->

## Related issue number

<!-- For example: "Closes #1234" -->

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

jjyao commented on python/ray/autoscaler/v2/instance_manager/node_provider.py:
```diff
@@ -184,124 +240,365 @@ def poll_errors(self) -> List[CloudInstanceProviderError]:
+
+        # Queues to retrieve new errors occur in the multi-thread executors
+        # temporarily.
+        self._launch_q = Queue()
```
_launch_queue

rickyyx commented on python/ray/tests/autoscaler_test_utils.py:
```diff
@@ -2,12 +2,14 @@
 
 class MockNode:
-    def __init__(self, node_id, tags, node_config, node_type, unique_ips=False):
+    def __init__(
```
added a few fields for controlling its behaviour at tests.

jjyao commented on python/ray/autoscaler/v2/instance_manager/node_provider.py:
```diff
@@ -64,14 +77,54 @@ class LaunchNodeError(CloudInstanceProviderError):
+    def __init__(
+        self,
+        node_type: NodeType,
+        count: int,
```
Can it be partial failure?

jjyao commented on python/ray/autoscaler/v2/instance_manager/node_provider.py:
```diff
@@ -184,124 +241,247 @@ def poll_errors(self) -> List[CloudInstanceProviderError]:
+        # Check node type is valid.
+        try:
+            if node_type not in config.get_node_type_configs():
+                raise ValueError(f"Invalid node type {node_type}")
```
This can be thrown earlier?

jjyao replied:
A single cycle uses the same config.

rickyyx commented on python/ray/autoscaler/_private/fake_multi_node/node_provider.py:
```diff
@@ -239,21 +239,40 @@ class FakeMultiNodeProvider(NodeProvider):
+        # Whether to launch multiple nodes at once, or one by one regardless of
+        # the count (default)
+        self._launch_multiple = provider_config.get("launch_multiple", False)
+        self._error_creates = None
```
yep. i actually thought this fake mult node provider is for testing itself? 

but i can call it out.

iycheng replied:
i think it's only used in testing

iycheng commented:
I feel good about the PR. But not too involved in this area. I'll let others approve.

jjyao commented on python/ray/autoscaler/_private/fake_multi_node/node_provider.py:
```diff
@@ -239,21 +239,40 @@ class FakeMultiNodeProvider(NodeProvider):
+        # Whether to launch multiple nodes at once, or one by one regardless of
+        # the count (default)
+        self._launch_multiple = provider_config.get("launch_multiple", False)
+        self._error_creates = None
```
Add a comment, maybe consider renaming it to `creation_error`?

rickyyx commented on python/ray/autoscaler/v2/tests/test_node_provider.py:
```diff
@@ -1,97 +1,306 @@
+        config = self.config_reader.get_autoscaling_config()
+        self.ray_session = None
+
+        if self.kind == "mock":
```
this is how I imagine we would add more node providers to the unit testing when we start supporting all of them (with mock client like aws client, or k8s client)

rickyyx commented on python/ray/autoscaler/v2/tests/test_node_provider.py:
```diff
@@ -1,97 +1,326 @@
+@pytest.fixture(scope="function")
+def get_provider():
+    def _get_provider(name, **kwargs):
+        if name == "fake_multi":
```
we should be adding different `XXXTester` for different providers.

jjyao commented on python/ray/autoscaler/_private/fake_multi_node/node_provider.py:
```diff
@@ -359,6 +407,18 @@ def _terminate_node(self, node):
+    def _test_add_creation_errors(self, e: Exception):
+        """Set an error that will be raised on
+        create_node_with_resources_and_labels."""
+        self._creation_errors = e
```
_creation_error

jjyao commented on python/ray/autoscaler/_private/fake_multi_node/node_provider.py:
```diff
@@ -359,6 +407,18 @@ def _terminate_node(self, node):
+    ############################
+    # Test only methods
+    ############################
+    def _test_add_creation_errors(self, e: Exception):
```
_test_set_

jjyao commented on python/ray/autoscaler/v2/schema.py:
```diff
@@ -19,6 +20,14 @@
+
+    HEAD = NODE_KIND_HEAD
+    WORKER = NODE_KIND_WORKER
+    UNMANAGED = NODE_KIND_UNMANAGED
```
Add comments

jjyao commented on python/ray/autoscaler/v2/tests/test_node_provider.py:
```diff
@@ -1,97 +1,326 @@
-            EventSummarizer(),
-            self.availability_tracker,
+
+class CloudProviderTesterBase(ICloudInstanceProvider):
```
CloudInstanceProviderTesterBase

jjyao commented on python/ray/autoscaler/v2/tests/test_node_provider.py:
```diff
@@ -1,97 +1,326 @@
+    ############################
+    # Test mock methods
+    ############################
+    def _add_creation_errors(self, e: Exception):
```
_set_creation_error

jjyao commented on python/ray/autoscaler/v2/instance_manager/node_provider.py:
```diff
@@ -184,124 +247,250 @@ def poll_errors(self) -> List[CloudInstanceProviderError]:
+
+        # Queues to retrieve new errors occur in the multi-thread executors
+        # temporarily.
+        self._launch_queue = Queue()
```
This is not used anymore?

rickyyx replied:
right - good catch.

jjyao reviewed (commented):
Lg

jjyao commented on python/ray/autoscaler/v2/schema.py:
```diff
@@ -19,6 +20,18 @@
+    # Worker node kind
+    WORKER = NODE_KIND_WORKER
+    # Unmanaged node kind (e.g. user started cloud instance or non
+    # ray node, this happens in KubeRay)
```
what's non ray node?

rickyyx replied:
changed comments - non-ray node should not be returned by node provider.

rickyyx commented:
- moved `NodeKind` to protobuf for usage from other places. 
- added missing support of `fake_multiple` node provider imports and v2 compatible flags.
